### PR TITLE
NFC TNEP improvements

### PR DESCRIPTION
**Pull request by KAGA164**
This PR provides improvements for the NFC TNEP library which ensures allows to define NDEF records locally inside
callback function.

### CODE REVIEW

File: samples/nfc/tnep_tag/src/main.c [hunk 1]
```diff
@@ -33,12 +29,8 @@ static const uint8_t svc_two_uri[] = "svc:e";
...
 static uint8_t tag_buffer2[NDEF_TNEP_MSG_SIZE];
 
-NFC_NDEF_MSG_DEF(ndef_msg, 16);
+static const uint8_t en_code[] = {'e', 'n'};
 
-NFC_NDEF_TEXT_RECORD_DESC_DEF(nfc_text, UTF_8, en_code, sizeof(en_code),
-			      msg, sizeof(msg));
-NFC_NDEF_TEXT_RECORD_DESC_DEF(svc_one_rec, UTF_8, en_code, sizeof(en_code),
-			      svc_one_msg, sizeof(svc_one_msg));
 NFC_NDEF_TEXT_RECORD_DESC_DEF(svc_two_rec, UTF_8, en_code, sizeof(en_code),
```
Comment (kapi-no) [inline samples/nfc/tnep_tag/src/main.c:34]: Can we also make this record descriptor automatic?

File: subsys/nfc/tnep/tag.c [hunk 2]
```diff
@@ -598,20 +559,24 @@ int nfc_tnep_tag_init(struct k_poll_event *events, uint8_t event_cnt,
...
+		LOG_ERR("No callback provides when maximum NDEF Record count id greater than 0");
+		return -EINVAL;
+	}
+
 	tnep.svc = svc;
 	tnep.svc_cnt = svc_cnt;
-	tnep.records = records;
-	tnep.records_cnt = records_cnt;
+	tnep.max_records_cnt = max_record_cnt;
+	tnep.encode = msg_encode_cb;
```
Comment (kapi-no) [inline subsys/nfc/tnep/tag.c:579]:
```suggestion
	tnep.initial_msg_encode = msg_encode_cb;
```

File: samples/nfc/tnep_tag/src/main.c (hunk 1 above)
Comment (KAGA164) [reply samples/nfc/tnep_tag/src/main.c]: It can be I moved it to local scope

File: subsys/nfc/tnep/tag.c (hunk 2 above)
Comment (KAGA164) [reply subsys/nfc/tnep/tag.c]: Done

File: subsys/nfc/tnep/tag.ld [hunk 3]
```diff
@@ -0,0 +1 @@
+Z_ITERABLE_SECTION_ROM(nfc_tnep_tag_service, 4)
```
Comment (kapi-no) [inline subsys/nfc/tnep/tag.ld:1]: What does 4 mean here? Does it stand for the number of section items that you can add (max 4 TNEP services)?

If so, please check if KConfig symbols are visible here and define this value as a parameter

File: include/nfc/tnep/tag.h [hunk 4]
```diff
@@ -172,29 +200,48 @@ int nfc_tnep_tag_tx_msg_buffer_register(uint8_t *tx_buff,
...
+/**
+ * @brief Encode the Initial NDEF Message
+ *
+ * This function encodes the Initial TNEP NDEF Message. It must be used
+ * in combination with @ref initial_msg_encode_t callback function.
+ *
+ * @param[in] msg Pointer to the Initial NDEF Message descriptor. NDEF Records
+ *                can be added to it also before calling this function.
+ * @param[in] records Pointer to the first NDEF Records structure. Can be set
+ *                    to NULL if there are not additional NDEF Records to
```
Comment (grochu) [inline include/nfc/tnep/tag.h:233]:
```suggestion
 *                    to NULL if there are no additional NDEF Records to
```

File: include/nfc/tnep/tag.h [hunk 5]
```diff
@@ -230,15 +277,19 @@ void nfc_tnep_tag_rx_msg_indicate(const uint8_t *rx_buffer, size_t len);
  * To indicate that application has no more data use
  * nfc_tnep_tag_tx_msg_no_app_data().
  *
- * @param[in] record Pointer to application data records.
- * @param[in] cnt Records count.
+ * @param[in] msg Pointer to NDEF Message with application data. The Message
+ *                must have at least one free place for the TNEP Status Record
```
Comment (grochu) [inline include/nfc/tnep/tag.h:281]:
```suggestion
 *                must have at least one free slot for the TNEP Status Record
```

File: samples/nfc/tnep_tag/src/main.c [hunk 6]
```diff
@@ -149,15 +144,38 @@ static void nfc_callback(void *context, enum nfc_t4t_event event,
...
 }
 
+/** .. include_startingpoint_initial_msg_cb_rst */
+static int tnep_initial_msg_encode(struct nfc_ndef_msg_desc *msg)
+{
+	const uint8_t text[] = "Hello World";
+
+	NFC_NDEF_TEXT_RECORD_DESC_DEF(nfc_text, UTF_8, en_code,
+				      sizeof(en_code), text,
+				      sizeof(text));
```
Comment (grochu) [inline samples/nfc/tnep_tag/src/main.c:154]: Would be better to use `strlen` to exclude the null terminator.

File: subsys/nfc/tnep/tag.c [hunk 7]
```diff
@@ -182,25 +184,26 @@ static int tnep_tx_initial_msg_set(void)
...
-	if (tnep.records && tnep.records_cnt) {
-		for (size_t i = 0; i < tnep.records_cnt; i++) {
-			err = tnep_tx_msg_add_rec(&tnep.records[i]);
-			if (err) {
-				return err;
-			}
-		}
+	tnep_tx_msg_clear(&NFC_NDEF_MSG(initial_msg));
+
+	/* If initial message callback is pass call it to encode
```
Comment (grochu) [inline subsys/nfc/tnep/tag.c:191]:
```suggestion
	/* If initial message callback is available call it to encode
```

File: subsys/nfc/tnep/tag.c [hunk 8]
```diff
@@ -182,25 +184,26 @@ static int tnep_tx_initial_msg_set(void)
...
+	 * no TNEP NDEF Records.
```
Comment (grochu) [inline subsys/nfc/tnep/tag.c:192]:
```suggestion
	 * non-TNEP NDEF Records.
```

Comment (KAGA164) [reply subsys/nfc/tnep/tag.c]: Done

File: subsys/nfc/tnep/tag.c (hunk 7 above)
Comment (KAGA164) [reply subsys/nfc/tnep/tag.c]: Done

File: samples/nfc/tnep_tag/src/main.c (hunk 6 above)
Comment (KAGA164) [reply samples/nfc/tnep_tag/src/main.c]: Ok changed in all places

File: include/nfc/tnep/tag.h (hunk 5 above)
Comment (KAGA164) [reply include/nfc/tnep/tag.h]: Done

File: include/nfc/tnep/tag.h (hunk 4 above)
Comment (KAGA164) [reply include/nfc/tnep/tag.h]: Done

File: subsys/nfc/tnep/tag.c [hunk 9]
```diff
@@ -16,11 +16,18 @@ LOG_MODULE_REGISTER(nfc_tnep_tag, CONFIG_NFC_TNEP_TAG_LOG_LEVEL);
 #define TNEP_EVENT_RX_IDX 0
 #define TNEP_EVENT_TX_IDX 1
 
+#define SERVICE_COUNT_GET(_count)                           \
+	extern struct nfc_tnep_tag_service                  \
+			_nfc_tnep_tag_service_list_start[]; \
+	extern struct nfc_tnep_tag_service                  \
+			_nfc_tnep_tag_service_list_end[];   \
```
[the remark this answers is not in the document]
Comment (KAGA164) [reply subsys/nfc/tnep/tag.c]: I removed this macro and declared this extern variables in function scope. I think we do not need it in the file scope.

File: include/nfc/tnep/tag.h [hunk 10]
```diff
@@ -25,6 +26,19 @@
...
+ * @brief Macro for creating and initializing an NFC NDEF TNEP
+ *        Application message descriptor.
+ *
+ * Use the macro @ref NFC_NDEF_MSG to access the NDEF message descriptor
+ * instance. This macro reserves place for the TNEP Status Record.
+ *
+ * @param name Name of the related instance.
+ * @param max_record_cnt Maximal count of records in the message.
+ */
+#define NFC_TNEP_TAG_APP_MSG(_name, _max_record_cnt) \
```
[the remark this answers is not in the document]
Comment (KAGA164) [reply include/nfc/tnep/tag.h]: Done

File: include/nfc/tnep/tag.rst [hunk 11]
```diff
@@ -35,34 +35,47 @@ The client (application or upper-layer protocol) must define these services and
...
-The service table is given to the :cpp:func:`nfc_tnep_tag_initial_msg_create` function and the Service Parameter record of each service is inserted into the initial TNEP message.
+The services data are loaded when calling the :cpp:func:`nfc_tnep_tag_initial_msg_create` function and the Service Parameter record of each service is inserted into the initial TNEP message.
 A Service Parameter record contains parameters for communicating with the service.
+The additional NDEF Records can be also encoded into the Initial TNEP message, as follows:
 
-See the following code for an example of service definition and TNEP initialization:
+* Pass a callback to :cpp:func:`nfc_tnep_tag_initial_msg_create`.
+* Next, define an additional Record inside a callback and call :cpp:func:`nfc_tnep_initial_msg_encode`
+
+See the following code for an example of the Initial Message creation:
```
Comment (b-gent) [inline include/nfc/tnep/tag.rst:50]: What is the difference here?
See the following code for an example of the Initial Message creation (this line)
vs 
The following code shows how to create the Initial NDEF Message (line 57)

File: include/nfc/tnep/tag.h [hunk 12]
```diff
@@ -25,6 +26,19 @@
 /** Maximum Waiting Time extension. */
 #define NFC_TNEP_TAG_MAX_N_WAIT_TIME 15
 
+/**
+ * @brief Macro for creating and initializing an NFC NDEF TNEP
```
[the remark this answers is not in the document]
Comment (FrancescoSer) [reply include/nfc/tnep/tag.h]: Fixed (considering line 31 too)

File: include/nfc/tnep/tag.h [hunk 13]
```diff
@@ -172,29 +200,48 @@ int nfc_tnep_tag_tx_msg_buffer_register(uint8_t *tx_buff,
  *           Otherwise, a (negative) error code is returned.
  */
 int nfc_tnep_tag_init(struct k_poll_event *events, uint8_t event_cnt,
-		      struct nfc_ndef_msg_desc *msg,
 		      nfc_payload_set_t payload_set);
 
 /**
- * @brief Create the initial TNEP NDEF message.
+ * @brief Create the Initial TNEP NDEF Message.
```
[the remark this answers is not in the document]
Comment (FrancescoSer) [reply include/nfc/tnep/tag.h]: Fixed in tag.h and tag.rst

File: include/nfc/tnep/tag.h [hunk 14]
```diff
@@ -25,6 +26,19 @@
...
 
+/**
+ * @brief Macro for creating and initializing an NFC NDEF TNEP
+ *        Application message descriptor.
+ *
+ * Use the macro @ref NFC_NDEF_MSG to access the NDEF message descriptor
+ * instance. This macro reserves place for the TNEP Status Record.
+ *
+ * @param _name Name of the related instance.
+ * @param _max_record_cnt Maximal count of records in the message.
```
[the remark this answers is not in the document]
Comment (FrancescoSer) [reply include/nfc/tnep/tag.h]: fixed

File: include/nfc/tnep/tag.h [hunk 15]
```diff
@@ -172,29 +200,48 @@ int nfc_tnep_tag_tx_msg_buffer_register(uint8_t *tx_buff,
...
- * @brief Create the initial TNEP NDEF message.
+ * @brief Create the Initial TNEP NDEF Message.
  *
- * This function create the Initial TNEP message. Initial NDEF message
- * has to contain at least one service parameters record. It can
- * contain also optional NDEF Records which can be used by NFC Poller
- * Device which does not support TNEP Protocol.
+ * This function creates the Initial TNEP message. Initial NDEF message
+ * has to contain at least one service parameters record defined
+ * using @ref NFC_TNEP_TAG_SERVICE_DEF. It can contain also optional NDEF
```
[the remark this answers is not in the document]
Comment (FrancescoSer) [reply include/nfc/tnep/tag.h]: fixed

File: include/nfc/tnep/tag.h [hunk 16]
```diff
@@ -230,15 +277,19 @@ void nfc_tnep_tag_rx_msg_indicate(const uint8_t *rx_buffer, size_t len);
  * To indicate that application has no more data use
  * nfc_tnep_tag_tx_msg_no_app_data().
  *
- * @param[in] record Pointer to application data records.
- * @param[in] cnt Records count.
+ * @param[in] msg Pointer to NDEF Message with application data. The Message
```
[the remark this answers is not in the document]
Comment (FrancescoSer) [reply include/nfc/tnep/tag.h]: fixed

File: include/nfc/tnep/tag.rst [hunk 17]
```diff
@@ -35,34 +35,47 @@ The client (application or upper-layer protocol) must define these services and
...
 To create a TNEP service, use the :c:macro:`NFC_TNEP_TAG_SERVICE_DEF` macro.
+The service data is stored in the persistent storage.
 
-After defining the services, create a table that lists all of them.
-You can use :c:macro:`NFC_TNEP_TAG_SERVICE` to get the names of the created services.
+Initial TNEP NDEF message
+=========================
 
-The service table is given to the :cpp:func:`nfc_tnep_tag_initial_msg_create` function and the Service Parameter record of each service is inserted into the initial TNEP message.
+The services data are loaded when calling the :cpp:func:`nfc_tnep_tag_initial_msg_create` function and the Service Parameter record of each service is inserted into the initial TNEP message.
```
[the remark this answers is not in the document]
Comment (FrancescoSer) [reply include/nfc/tnep/tag.rst]: fixed

File: include/nfc/tnep/tag.rst [hunk 18]
```diff
@@ -35,34 +35,47 @@ The client (application or upper-layer protocol) must define these services and
...
 A Service Parameter record contains parameters for communicating with the service.
+The additional NDEF Records can be also encoded into the Initial TNEP message, as follows:
```
[the remark this answers is not in the document]
Comment (FrancescoSer) [reply include/nfc/tnep/tag.rst]: fixed in tag.rst

File: include/nfc/tnep/tag.rst [hunk 19]
```diff
@@ -35,34 +35,47 @@ The client (application or upper-layer protocol) must define these services and
...
 
-See the following code for an example of service definition and TNEP initialization:
+* Pass a callback to :cpp:func:`nfc_tnep_tag_initial_msg_create`.
```
[the remark this answers is not in the document]
Comment (FrancescoSer) [reply include/nfc/tnep/tag.rst]: fixed

File: include/nfc/tnep/tag.rst [hunk 20]
```diff
@@ -35,34 +35,47 @@ The client (application or upper-layer protocol) must define these services and
...
+* Next, define an additional Record inside a callback and call :cpp:func:`nfc_tnep_initial_msg_encode`
```
[the remark this answers is not in the document]
Comment (FrancescoSer) [reply include/nfc/tnep/tag.rst]: fixed

File: include/nfc/tnep/tag.rst (hunk 11 above)
Comment (FrancescoSer) [reply include/nfc/tnep/tag.rst]: Not fixed yet, some additional clarification are needed. Anyway, the idea was to not have 2 different blocks of code one after the other without any text description between them 
@KAGA164 FYA

Comment (KAGA164) [reply include/nfc/tnep/tag.rst]: The first sentence shall say encoding not creation. I fixed it

**Comment by KAGA164** [comment]
@b-gent @FrancescoSer Are we ready to merge ?

File: include/nfc/tnep/tag.h [hunk 21]
```diff
@@ -172,29 +200,48 @@ int nfc_tnep_tag_tx_msg_buffer_register(uint8_t *tx_buff,
...
+ * records which can be used by NFC Poller Device which does not support
+ * TNEP Protocol.
  *
- * @param[in] svc Pointer to the first service information structure.
- * @param[in] svc_cnt Number of provided services for application.
- * @param[in] records Pointer to the first no TNEP NDEF Records structure.
- * @param[in] records_cnt Number of privided NDEF Records.
+ * @param[in] max_record_cnt Maximum count of the optional NDEF records
+ * @param[in] msg_encode_cb Callback function for encoding the
+ *                          Initial TNEP NDEF message.
```
Comment (doki-nordic) [inline include/nfc/tnep/tag.h:216]: I see in examples that this parameter is optional and can be NULL. This should be clearly stated in parameter documentation.

Comment (KAGA164) [reply include/nfc/tnep/tag.h]: Ok. I added an additional description for this parameters.

**Review by anangl — APPROVED** [review]

**Pull request merged**